### PR TITLE
cosmos events parser

### PR DESCRIPTION
**Pull request by agouin**
Stub out `CosmosChainProcessor`

Adds code to parse IBC messages from `*abci.ResponseDeliverTx`, which is returned by tendermint's `BlockResults` query.

Introduces types used for parsing messages, later to be used for constructing IBC messages within the `CosmosChainProcessor` before passing to the `PathProcessor`

`ibcMessage`
`channelInfo`
`connectionInfo`
`packetInfo`
`clientInfo`

These types are private because they are only used within the `CosmosChainProcessor`

### CODE REVIEW

File: relayer/chains/cosmos/types.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+type TransactionMessage struct {
+	Action      string
+	PacketInfo  *PacketInfo
+	ChannelInfo *ChannelInfo
+	ClientInfo  *ClientInfo
+	// TODO ConnectionInfo
+}
+
+// ChannelInfo contains useful channel information during channel state changes
+type ChannelInfo struct {
```
Comment (jackzampolin) [inline relayer/chains/cosmos/types.go]: Do we also want a connection info type?

File: relayer/chains/cosmos/types.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	PortID                   string
+	ChannelID                string
+	CounterpartyClientID     string
+	CounterpartyConnectionID string
+	CounterpartyPortID       string
+	CounterpartyChannelID    string
+}
+
+// PacketInfo contains pertinent packet information for constructing IBC messages for the counterparty.
+type PacketInfo struct {
```
Comment (jackzampolin) [inline relayer/chains/cosmos/types.go]: Should we separate this into a Packet and Acknowledgement struct?

File: relayer/chains/cosmos/types.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package cosmos
+
+import (
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	chantypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+)
+
+// TransactionMessage is the type used for parsing IBC messages
+type TransactionMessage struct {
```
Comment (jackzampolin) [inline relayer/chains/cosmos/types.go]: Should this be an interface instead of a struct?

**Review by jackzampolin — COMMENTED** [review]
Good start, some questions about the structs

File: relayer/chains/cosmos/types.go (hunk 1 above)
Comment (agouin) [reply relayer/chains/cosmos/types.go]: added

File: relayer/chains/cosmos/types.go (hunk 3 above)
Comment (agouin) [reply relayer/chains/cosmos/types.go]: This is an internal intermediate structure for capturing all possible fields from ibc messages. I updated these types and the properties within them to private.

 It is used to pass all possible fields to the ibc message handlers, which will also be internal to the `CosmosChainProcessor` and will come in a future PR (implemented in POC [here](https://github.com/cosmos/relayer/pull/743/files#diff-93e4956f6ae840c202f7e921be7e87850093b28413799f5a09fac8e21442ec09)). The ibc message handlers can take the relevant properties based on the `messageType` to assemble a `provider.RelayerMessage` that will have the required properties to assemble the next message in the sequence for the counterparty chain. 

For example, if we parse a `MsgTransfer` here, `packetInfo` will be non-nil, then the ibc message handlers can assemble a `provider.RelayerMessage` with the required fields to construct a `MsgRecvPacket` for the counterparty chain.

File: relayer/chains/cosmos/types.go (hunk 2 above)
Comment (agouin) [reply relayer/chains/cosmos/types.go]: similar to the above, I was thinking this could be the type for all possible packet properties parsed from the events. the ibc message handlers in a future PR can construct the more concrete types before anything leaves the scope of the `CosmosChainProcessor`.

File: relayer/chains/cosmos/types.go (hunk 3 above)
Comment (agouin) [reply relayer/chains/cosmos/types.go]: I changed this to have a single `messageInfo interface{}` property, since, depending on the `messageType`, only one of the `Info` properties will be needed.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	pathProcessors processor.PathProcessors
+	ChainProvider  *cosmos.CosmosProvider
+}
+
+func NewCosmosChainProcessor(ctx context.Context, log *zap.Logger, rpcAddress string, provider provider.ChainProvider, pathProcessors processor.PathProcessors) (*CosmosChainProcessor, error) {
+	cosmosProvider, ok := provider.(*cosmos.CosmosProvider)
+	if !ok {
+		return nil, errors.New("only cosmos providers are supported for cosmos chain processors")
+	}
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: A panic should be fine here, as this indicates a programming error, not a recoverable runtime error.

File: relayer/chains/cosmos/event_parser.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	"go.uber.org/zap"
+)
+
+// processTransaction parses all events within a transaction to find IBC messages
+func (ccp *CosmosChainProcessor) processTransaction(tx *abci.ResponseDeliverTx) []ibcMessage {
+	messages := []ibcMessage{}
+	parsedLogs, err := sdk.ParseABCILogs(tx.Log)
+	if err != nil {
+		return messages
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: `ccp.log.Info("Failed to parse abci logs", zap.Error(err)); return nil` here?

File: relayer/chains/cosmos/event_parser.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+		for _, event := range messageLog.Events {
+			switch event.Type {
+			case "message":
+				for _, attr := range event.Attributes {
+					switch attr.Key {
+					case "action":
+						messageType = attr.Value
+					}
+				}
+			case "create_client", "update_client", "upgrade_client", "submit_misbehaviour":
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: Better if you can use existing constants instead of string literals for these. I assume all of them are declared in ibc-go.

File: relayer/chains/cosmos/event_parser.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+				"channel_open_ack", "channel_open_confirm", "channel_close_init", "channel_close_confirm":
+				messageInfo = ccp.parseChannelInfo(event.Attributes)
+			}
+		}
+		if messageInfo == nil {
+			// Not an IBC message, don't need to log here
+			continue
+		}
+		if messageType == "" {
+			ccp.log.Error("unexpected ibc message parser state: message info is populated but type is empty",
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: Log messages begin with an uppercase letter. https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/cosmos/event_parser.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+				RevisionNumber: revisionNumber,
+				RevisionHeight: revisionHeight,
+			}
+		}
+	}
+	return res
+}
+
+func (ccp *CosmosChainProcessor) parsePacketInfo(attributes []sdk.Attribute) packetInfo {
+	res := packetInfo{packet: chantypes.Packet{}}
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: No need to use initialize the `packet` field with a zero-value here.

My style preference is to use `var res packetInfo` in favor of `res := packetInfo{}` for these. I read the former as more of "I am explicitly not setting any fields" and the latter is more ambiguous about whether the author intended to set any fields. Likewise for `x := new(Foo)` vs. `x := &Foo{}`. It's a subtle distinction but I find it helpful.

File: relayer/chains/cosmos/event_parser.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+					zap.Error(err),
+				)
+				continue
+			}
+		case "packet_data":
+			res.packet.Data = []byte(attr.Value)
+		case "packet_data_hex":
+			data, err := hex.DecodeString(attr.Value)
+			if err == nil {
+				ccp.log.Error("error parsing packet data",
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: Should log and continue on `err != nil`.

File: relayer/chains/cosmos/event_parser.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+					zap.Error(err),
+				)
+				res.packet.Data = data
+			}
+		case "packet_ack":
+			res.ack = []byte(attr.Value)
+		case "packet_ack_hex":
+			data, err := hex.DecodeString(attr.Value)
+			if err == nil {
+				ccp.log.Error("error parsing packet ack",
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser.go]: Should log and continue on `err != nil`.

**Review by mark-rushakoff — COMMENTED** [review]
There are two error checks that appear to have mistakes.

Style-wise, I would prefer to see existing constants from the appropriate external packages for the event type switches, and our logging doc specifies that log messages begin with uppercase letters. It is probably worth dropping the error messages to Info level if there is nothing the operator can do to prevent malformed incoming packets.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 4 above)
Comment (DavidNix) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: Why can't we accept the concrete type `*cosmos.CosmosProvider` as an argument vs `provider.ChainProvider`?

File: relayer/chains/cosmos/event_parser.go [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+// processTransaction parses all events within a transaction to find IBC messages
+func (ccp *CosmosChainProcessor) processTransaction(tx *abci.ResponseDeliverTx) []ibcMessage {
+	messages := []ibcMessage{}
+	parsedLogs, err := sdk.ParseABCILogs(tx.Log)
+	if err != nil {
+		return messages
+	}
+	for _, messageLog := range parsedLogs {
+		var messageInfo interface{}
```
Comment (DavidNix) [inline relayer/chains/cosmos/event_parser.go:28]: nit (style):

Because relayer is on go 1.18. I strongly prefer `any` to the more verbose `interface{}`. They are the same, but `any` reads a lot better. 

This refactor can be done in 1 command:

```
gofmt -w -r 'interface{} -> any' .
```

I wouldn't recommend that change in this PR as it will probably cause a lot of changes to sift through.

File: relayer/chains/cosmos/event_parser.go (hunk 9 above)
Comment (DavidNix) [reply relayer/chains/cosmos/event_parser.go]: I feel we use `continue` and for loops too much here and elsewhere. It lends itself to this exact problem, forgetting to continue. 

I think there's a pattern where we can abstract these parts into functions/methods with `return` values. That way, the compiler forces us to remember to return something. The compiler does not enforce continue in this manner.

E.g. 

```go
func parseValue(attr sdk.Attribute, packet *packetInfo) error {
  switch attr.Key {
  case "packet_sequence":
    packet.Sequence, err := strconv.ParseUint(attr.Value, 10, 64)
    if err != nil {
      return fmt.Errorf("parse sequence: %w", err)
    // ...
}

for _, attr := range attributes {
   parseValue(attr.Key, &info)
   // ...
```

Another option:

It may be less performant, but ibctest use a helper function which abstracts the for loop for this messy, loose event/attribute data structure. The calling code is much cleaner imo. https://github.com/strangelove-ventures/ibctest/blob/60cd4d1fb280e3e1d39e0cc77612e200073ef8e3/chain/cosmos/cosmos_chain.go#L152

**Review by DavidNix — COMMENTED** [review]
Straightforward!

On the whole, I think what's missing are unit tests around constructing these messages. It seems vital we are parsing the tendermint events appropriately. And important to test for error/edge cases.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 4 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: > Why can't we accept the concrete type *cosmos.CosmosProvider as an argument vs provider.ChainProvider?

I was thinking it might make it easier for the caller so they don't have to cast, but I'm fine with making the API explicit instead of enforcing it here.

File: relayer/chains/cosmos/event_parser.go (hunk 5 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: done

File: relayer/chains/cosmos/event_parser.go (hunk 6 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: done

File: relayer/chains/cosmos/event_parser.go (hunk 7 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: done

File: relayer/chains/cosmos/event_parser.go (hunk 8 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: done

File: relayer/chains/cosmos/event_parser.go (hunk 9 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go]: > Should log and continue on err != nil.

good catch.

>I think there's a pattern where we can abstract these parts into functions/methods with return values.

done. I think I prefer the return pattern over unnecessary iterations here for performance reasons, but I do like how nice the API is for ibctest.

File: relayer/chains/cosmos/event_parser.go (hunk 11 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser.go:28]: I had updated it to go 1.18 in my POC PR, but main is not on 1.18 yet. I'm happy to handle this in a follow up though.

**Comment by agouin** [comment]
> On the whole, I think what's missing are unit tests around constructing these messages.

Unit tests added!

File: relayer/chains/cosmos/event_parser.go (hunk 11 above)
Comment (jackzampolin) [reply relayer/chains/cosmos/event_parser.go:28]: TIL, and holy f this is amazing

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+
+	"github.com/cosmos/relayer/v2/relayer/processor"
+	"github.com/cosmos/relayer/v2/relayer/provider/cosmos"
+	"go.uber.org/zap"
+)
+
+type CosmosChainProcessor struct {
+	ctx context.Context
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: Looks like this context is not referenced anywhere -- which is good since [the docs for context state](https://pkg.go.dev/context):

> Do not store Contexts inside a struct type

The field can be removed and the parameter can be removed from `NewCosmosChainProcessor`.

File: relayer/chains/cosmos/cosmos_chain_processor.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+	log *zap.Logger
+
+	pathProcessors processor.PathProcessors
+	ChainProvider  *cosmos.CosmosProvider
+}
+
+func NewCosmosChainProcessor(ctx context.Context, log *zap.Logger, rpcAddress string, provider *cosmos.CosmosProvider, pathProcessors processor.PathProcessors) (*CosmosChainProcessor, error) {
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/cosmos_chain_processor.go]: `ctx` can be removed, and based on the usage here, it should be safe to change the signature to stop returning an error.

**Review by mark-rushakoff — COMMENTED** [review]
I have to break now and won't be able to return to this review for a couple hours, so just leaving a couple more comments.

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 12 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: done

File: relayer/chains/cosmos/cosmos_chain_processor.go (hunk 13 above)
Comment (agouin) [reply relayer/chains/cosmos/cosmos_chain_processor.go]: yes, the error is for future additions, but removed for now.

File: relayer/chains/cosmos/event_parser_test.go [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+
+	require.Equal(t, uint64(1), parsed.packet.TimeoutHeight.RevisionNumber, "packet timeout height revision number does not match")
+	require.Equal(t, uint64(1245), parsed.packet.TimeoutHeight.RevisionHeight, "packet timeout height revision height does not match")
+
+	require.Equal(t, uint64(1654033235600000000), parsed.packet.TimeoutTimestamp, "packet timeout timestamp does not match")
+
+	require.Equal(t, testPacketSrcChannel, parsed.packet.SourceChannel, "packet source channel does not match")
+	require.Equal(t, testPacketSrcPort, parsed.packet.SourcePort, "packet source port does not match")
+	require.Equal(t, testPacketDstChannel, parsed.packet.DestinationChannel, "packet destination channel does not match")
+	require.Equal(t, testPacketDstPort, parsed.packet.DestinationPort, "packet destination port does not match")
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser_test.go]: Once a struct gets this many fields, I often reach to [go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) to simplify comparison tests like this:

```go
parsed := ...

require.Empty(t, cmp.Diff(parsed, packetInfo{
  packet: packet{
    Sequence: 1,
    // ...
  },
})
```

It is easy to explicitly configure cmp.Diff to ignore one particular field, or to otherwise set custom comparison rules. 

But if there is a good reason to keep it the way it is, that's fine.

File: relayer/chains/cosmos/event_parser_test.go [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+
+func TestParsePacket(t *testing.T) {
+	testPacketTimeoutHeight := "1-1245"
+	testPacketTimeoutTimestamp := "1654033235600000000"
+	testPacketSequence := "1"
+	testPacketDataHex := "0123456789ABCDEF"
+	testPacketSrcChannel := "channel-0"
+	testPacketSrcPort := "port-0"
+	testPacketDstChannel := "channel-1"
+	testPacketDstPort := "port-1"
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser_test.go]: Stylistically, I would rather see this as a `const` block to be more explicit about not allowing reassignment (and in the other new test functions too).

File: relayer/chains/cosmos/event_parser_test.go [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+
+	ibcMessages := parseABCILogs(zap.NewNop(), testChainID1, abciLogs)
+
+	require.Equal(t, len(ibcMessages), 2)
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/event_parser_test.go]: Prefer `require.Len(t, ibcMessages, 2)` which will produce a clearer default error message.

**Review by mark-rushakoff — APPROVED** [review]
I left a couple more minor style comments, but nothing blocking a merge.

File: relayer/chains/cosmos/event_parser_test.go (hunk 14 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser_test.go]: looks great. done

File: relayer/chains/cosmos/event_parser_test.go (hunk 16 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser_test.go]: done

File: relayer/chains/cosmos/event_parser_test.go (hunk 15 above)
Comment (agouin) [reply relayer/chains/cosmos/event_parser_test.go]: done
